[GStreamer][1.26] Critical warnings in volume element

philn opened the pull request:
#### 97a032c9d8502c4c3250642ac7031785767b6085
<pre>
[GStreamer][1.26] Critical warnings in volume element
<a href="https://bugs.webkit.org/show_bug.cgi?id=290537">https://bugs.webkit.org/show_bug.cgi?id=290537</a>

Reviewed by Xabier Rodriguez-Calvar.

The seeks performed by oggdemux to determine the stream duration led our http source element to an
inconsistent state where haveSize was set to false and size to a valid value, so the
webKitWebSrcGetSize() call was failing even though we had a valid size. By removing haveSize and
changing size to an optional&lt;uint64_t&gt; the code is now less error-prone and easier to follow.

* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
(webkitWebSrcReset):
(restartLoaderIfNeeded):
(stopLoaderIfNeeded):
(webKitWebSrcCreate):
(webKitWebSrcGetSize):
(webKitWebSrcDoSeek):
(CachedResourceStreamingClient::responseReceived):
(CachedResourceStreamingClient::recalculateLengthAndSeekableIfNeeded):
(CachedResourceStreamingClient::dataReceived):

Canonical link: <a href="https://commits.webkit.org/294419@main">https://commits.webkit.org/294419@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/410821ce02cacabd545dc201b8a23a4c97a1e2bb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101768 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21436 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11752 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106926 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52402 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103808 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21744 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29940 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/77484 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/34510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104775 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16800 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/91901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57822 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16625 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/9919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51753 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86481 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/9996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109282 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28901 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/21283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86464 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29262 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/88102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86035 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21891 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30791 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/8512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23066 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28829 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34119 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28640 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31963 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30199 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
```diff
@@ -433,21 +430,21 @@ static void stopLoaderIfNeeded([[maybe_unused]] WebKitWebSrc* src, DataMutexLock
-    GST_TRACE_OBJECT(src, "queue size %zu (max %1.0f)", queueSize, members->size * HIGH_QUEUE_FACTOR_THRESHOLD);
-    if (queueSize <= members->size * HIGH_QUEUE_FACTOR_THRESHOLD) {
+    GST_TRACE_OBJECT(src, "queue size %zu (max %1.0f)", queueSize, *members->size * HIGH_QUEUE_FACTOR_THRESHOLD);
+    if (queueSize <= *members->size * HIGH_QUEUE_FACTOR_THRESHOLD) {
```
We need to protect against size being nullopt, maybe bailing out in the if above.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
```diff
@@ -433,21 +430,21 @@ static void stopLoaderIfNeeded([[maybe_unused]] WebKitWebSrc* src, DataMutexLock
     }
 
-    if (members->readPosition == members->size) {
+    if (members->readPosition == *members->size) {
```
If you didn't bail out in the if above, ditto.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
```diff
@@ -1141,18 +1135,17 @@ void CachedResourceStreamingClient::recalculateLengthAndSeekableIfNeeded(DataMut
         return;
 
-    if (members->haveSize || members->isSeekable || members->size)
+    if (members->isSeekable || members->size)
```
You are missing one check here, if size == 0.

philn replied:
Hmm? I don't understand, which size are you referring to?

calvaris replied:
The size you're turning into a std::optional. It can be nullopt, 0 or something different. This if used to bail out in the first two cases. With your change it would bail out only in the first, not in the second and third.

philn replied:
Ah yes, thanks.

philn committed:
[GStreamer][1.26] Critical warnings in volume element
https://bugs.webkit.org/show_bug.cgi?id=290537

Reviewed by Xabier Rodriguez-Calvar.

The seeks performed by oggdemux to determine the stream duration led our http source element to an
inconsistent state where haveSize was set to false and size to a valid value, so the
webKitWebSrcGetSize() call was failing even though we had a valid size. By removing haveSize and
changing size to an optional<uint64_t> the code is now less error-prone and easier to follow.

* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
(webkitWebSrcReset):
(restartLoaderIfNeeded):
(stopLoaderIfNeeded):
(webKitWebSrcCreate):
(webKitWebSrcGetSize):
(webKitWebSrcDoSeek):
(CachedResourceStreamingClient::responseReceived):
(CachedResourceStreamingClient::recalculateLengthAndSeekableIfNeeded):
(CachedResourceStreamingClient::dataReceived):

Canonical link: https://commits.webkit.org/294419@main
